AP_Common/AP_Avoidance: create AP_ExpandingArray class and use for avoidance

rmackay9 opened the pull request:
This PR allows the AP_Avoidance/AP_OADijkstra class to find the shortest path to a destination even if there are many more than 20 polygon fence points.  This is accomplished by making use of a new AP_ExpandingArray template class to hold various bits of required data including:

- the inner polygon fence points (the inner polygon fence is 10m within the regular polygon fence)
- the visibility graphs which hold the distances between all "nodes" (aka safe positions) which includes the waypoint origin, waypoint destination and the inner polygon fence points
- the resulting shortest path from origin to destination

This has been tested in SITL using a polygon fence of about 40 points (see image below) but probably needs a bit more testing to ensure we completely handle the case where we run out of memory.  Perhaps we should also add limits on how big the arrays can get.

It's possible that I've pushed this PR a little prematurely (i.e. I should do more testing) but in any case, all feedback is appreciated!

![rover-snake-fence](https://user-images.githubusercontent.com/1498098/59547760-fe808e80-8f7e-11e9-8a2d-534f727a5cc3.png)

This PR resolves this issue https://github.com/ArduPilot/ardupilot/issues/11572.

WickedShell commented on libraries/AC_Avoidance/AP_OAVisGraph.cpp:
```diff
@@ -24,27 +24,17 @@ AP_OAVisGraph::AP_OAVisGraph(uint8_t size)
     // check there is space
-    if (_num_items >= _num_items_max) {
-        return false;
+    if (_num_items >= _items.max_items()) {
```
In practice how is this any better then just calling `_items.expand_to_hold(_num_items + 1);` Seems like a more verbose version of that call, and one that is frequently needed.

rmackay9 replied:
sure, replaced with expand_to_hold. txs

WickedShell commented on libraries/AP_Common/AP_ExpandingArray.h:
```diff
@@ -0,0 +1,115 @@
+
+    /* Do not allow copies */
+    AP_ExpandingArray<T>(const AP_ExpandingArray<T> &other) = delete;
+    AP_ExpandingArray<T> &operator=(const AP_ExpandingArray<T>&) = delete;
```
We may actually want to implement delete for this, since otherwise this will leak memory. We don't normally delete things, but most things aren't manipulating memory the way this is.

rmackay9 replied:
I'm not sure I understand how we could be leaking memory.  The array may expand to hold everything required but this isn't technically the same as "leaking" which implies the memory is completely lost after use.

So far we don't have a need to delete these objects though so I think it's OK that these definitions that save developers from accidentally creating copies remain.

WickedShell replied:
The two operators you overload here block copies and assignment. They don't blocking deleting the object. If you ever allocate then delete one of these arrays all the backing memory the array allocates is forever lost.

rmackay9 replied:
@WickedShell, ah, I undestand.  OK, I'll add a destructor.

rmackay9 replied:
destructor added

rmackay9 commented:
@WickedShell, txs very much for the review.  I'll certainly check it with valgrind.
Re the name of the class and internal variables, naming is hard so any suggestions are very welcome!

rmackay9 commented on libraries/AP_Common/AP_ExpandingArray.h:
```diff
@@ -0,0 +1,115 @@
+
+    // allow use as an array for assigning to elements. no bounds checking is performed
+    T &operator[](uint16_t i)
+    {
```
I don't think range checking is required because we don't have it on regular arrays.  Instead I think we can handle the danger (which we already have for existing arrays) with developer education .. in particular I'll add some comments at the top of the .h file explaining how it works and how it should be used.

magicrub replied:
Uhm, we have range checks in many places. Also, there's several divide-by-zero checks that need to be added, especially since the default chunk_size is zero.

rmackay9 replied:
discussed on dev call and decided range checks weren't necessary for now at least

rmackay9 commented on libraries/AP_Common/AP_ExpandingArray.h:
```diff
@@ -0,0 +1,115 @@
+    }
+
+    // allow use as an array for accessing elements. no bounds checking is performed
+    const T &operator[](uint16_t i) const
```
same discussion as above so I'll mark this particular comment as resolved.

rmackay9 commented on libraries/AP_Common/AP_ExpandingArray.h:
```diff
@@ -0,0 +1,115 @@
+#include <AP_Common/AP_Common.h>
+
+template <typename T>
+class AP_ExpandingArray
```
I hope the improve documentation I'll add at the top will address the concerns about using memcpy.

I'm very happy to change the name but naming is hard and "ExpandingArray" is very easy to understand.  I'm struggling to think of something better so so I'd like some suggestions.

rmackay9 replied:
discussed on the dev call and decided the current name is as good as any alternatives

rmackay9 commented on libraries/AP_Common/AP_ExpandingArray.h:
```diff
@@ -0,0 +1,115 @@
+    // chunk_ptrs array is grown by this many elements each time it fills
+    const uint16_t chunk_ptr_increment = 50;
+
+    typedef T* chunk_ptr;
```
chunk_ptr is used up on line 66 as well.  I agree that it should at least be renamed to chunk_ptr_t.  I personally prefer having the typedef because otherwise it's easy to become confused with which of the degrees is being modified.  If people think it's more confusing I could remove the typedef and just use T*.

rmackay9 replied:
re renaming "chunk_size".. any suggestion on a better name?  "chunk_num_elements" perhaps?

rmackay9 replied:
we discussed this on the dev call and decided it was OK to leave the typedef in there.

rmackay9 commented on libraries/AP_Common/AP_ExpandingArray.h:
```diff
@@ -0,0 +1,115 @@
+        expand(1);
+    }
+
+    AP_ExpandingArray<T>(uint16_t num_chunks, uint16_t elements_per_chunk) :
```
I guess this is two separate suggestions.

- requiring the user to always specify the number of elements-per-chunk sounds fine to me.  I'll remove the default constructor.
- the idea for allowing the user to specify the number of chunks was to get the memory allocation out of the way earlier.  I.e. at startup before arming rather than at run-time.  That could be handled by the caller though pretty easily so I'll remove this.

rmackay9 commented on libraries/AP_Common/AP_ExpandingArray.h:
```diff
@@ -0,0 +1,115 @@
+            memcpy(chunk_ptrs_new, chunk_ptrs, chunk_count_max * sizeof(T*));
+
+            // free old pointers array
+            delete chunk_ptrs;
```
aha, I did not know this, txs!  replaced with "free()".

rmackay9 commented:
I've pushed two additional commits to resolve some of @WickedShell's feedback and also run it through Valgrind (no issues uncovered).  I've added DevCallTopic so we can perhaps make some decisions on some of the outstanding points we don't agree on but in any case, I very much appreciate the review.

magicrub commented on libraries/AP_Common/AP_ExpandingArray.h:
```diff
@@ -0,0 +1,128 @@
+    }
+
+    // expand the array by specified number of chunks, returns true on success
+    bool expand(uint16_t num_chunks = 1)
```
const num_chunks

rmackay9 replied:
I'm a little worried that this would send us down the path of putting "const" in front of all arguments.. it's really not necessary and I think it hurts readability so I think I might not make this change.

magicrub commented on libraries/AP_Common/AP_ExpandingArray.h:
```diff
@@ -0,0 +1,128 @@
+    {
+        // expand chunk_ptrs array if necessary
+        if (chunk_count + num_chunks >= chunk_count_max) {
+            uint16_t chunk_ptr_size = chunk_count + num_chunks + chunk_ptr_increment;
```
Why is chunk_ptr_increment 50? Why not 1?

rmackay9 replied:
I've reduced it to 32 .. if it was "1" then we would end up reallocating the chunk_ptrs array too often which could lead to fragmentation of memory.

magicrub commented:
Seems like we're at the point where we're better off using std::list (or std::vector) instead of re-inventing the wheel. Sure, we may be a little more lean but the risk of doing it wrong is a bad risk to take.

OXINARF commented:
I haven't done a review yet, but given this is likely to be discussed on the dev call before I'm able to finish, I have to agree with @magicrub and ask, in what ways is this better than just using a `std::vector`?

rmackay9 commented:
@magicrub, @OXINARF, thanks for the review, it's possible that std::vector could be used for the top level table of pointers (chunk_ptrs) but I think if we completely replaced this class with std::vector we would end up allocating and free-ing memory whenever the array must be expanded which would lead to memory fragmentation.

rmackay9 commented:
Feedback from the dev call:

- no need to bounds checking on the operator[] methods but we should add some checking for SITL (assert?)
- alternative names of ExpandingPool, ExpandingStore were proposed but team was generally OK with Randy's preference to stick with ExpandingArray
- chunk_size should be made constant and a factor of 2 to allow compiler to optimise the operator calcs use of %
- @tridge requested we attempt to move the majority of the implementation to a cpp file to keep flash size down if this is used in multiple places.  The RingBuffer class may be a good reference to look at
- realloc should be used instead of explicitly using calloc and free when resizing the chunk_ptrs array
- @tridge had no strong preference on whether the "typedef chunk_ptr_t T*" is kept or not

WickedShell commented on libraries/AP_Common/AP_ExpandingArray.h:
```diff
@@ -75,15 +76,10 @@ class AP_ExpandingArray
         if (chunk_count + num_chunks >= chunk_count_max) {
             uint16_t chunk_ptr_size = chunk_count + num_chunks + chunk_ptr_increment;
-            chunk_ptr_t *chunk_ptrs_new = (chunk_ptr_t*)calloc(chunk_ptr_size, sizeof(T*));
+            chunk_ptr_t *chunk_ptrs_new = (chunk_ptr_t*)realloc(chunk_ptrs, chunk_ptr_size * sizeof(T*));
```
I think it's safe here, but worth noting that before you were using calloc, which sets the new memory to 0, realloc won't do that for you. I think you have sufficient protection but I wanted to explicitly make sure you were tracking that.

rmackay9 replied:
@WickedShell, great, thanks for the warning, I had assume that like calloc/malloc it would set the memory to zero.  As you say, we always write before using so should be ok

rmackay9 commented:
I think I've addressed all issues from the reviews and those raised on the dev call including reducing the code size.  I've re-tested including using valgrind and it seems to work.

I'll give it a bit more time for feedback and then merge to master if there are no other comments or issues.  Thanks again for the reviews!

WickedShell commented:
> I'll give it a bit more time for feedback and then merge to master if there are no other comments or issues. Thanks again for the reviews!

Have you valgrind'd it?

rmackay9 commented:
@WickedShell, yes, it passed valgrind without issues.  Valgrind did seem to turn up an issue with memory access to do with executing missions (and perhaps changing a mission while it's executing it in Auto) but it happens on master as well and I will investigate it separately.

@tridge, I know you're busy but do you want to have one last look at how the implementations were moved to the .cpp file before I merge?

rmackay9 commented:
Merged, thanks!
